### PR TITLE
[reward, rollout, diffusion] feat: add DiNa-LRM support for SD3.5 FlowGRPO

### PR DESCRIPTION
**Pull request by GongyeLiu**
### What does this PR do?

> Add **concise** overview of what this PR aims to achieve or accomplish. Reference related GitHub issues and PRs that help with the review.

- Add [DiNa-LRM](https://github.com/HKUST-C4G/diffusion-rm) scoring for SD3.5 FlowGRPO.
- Add `image`, `latent`, and `both` rollout output modes.
- Add a PickScore-SFW data preprocessing script, an 8-GPU training recipe, and documentation.
- Build on the SD3.5 FlowGRPO support introduced in #178.

DiNa-LRM is a latent-space diffusion reward model. Unlike image reward models like HPS-v3, DiNa-LRM scores the clean diffusion latent directly. This results in faster rollout and lower GPU memory consumption.

The training rollout uses `output_type=latent` to skip VAE decoding, while validation uses `output_type=both` to preserve decoded images for logging.

The experiment uses PickScore-SFW prompts with:
- Base model: SD3.5-Medium
- Reward model: DiNa-LRM-SD35M-12layers
- 7 GPUs for actor and rollout workers
- 1 GPU for the DRM server

The hyperparameters used in the experiment are:  `clip_ratio=3e-6`, `lr=5e-5`, `rollout.n=18`, `kl_loss_coef=0.01`,  `train_batch_size=14`, `ppo_mini_batch_size=7`, `ppo_micro_batch_size_per_gpu=9`.
Namely, each training
step generates 252 rollout samples. The prompts are divided into two PPO mini-batches of 7 prompts each, resulting in two PPO update rounds per training step.

image

~100s to run one step (252 trajectories):
image

Peak VRAMs: 40G for 512x512 images:

image

Validation: ~80 steps is enough for improving quality:

image

image

### Checklist Before Starting

- [x] Search for similar PRs. Paste at least one query link here: ...
- [x] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `fsdp`, `vllm_omni`, `rollout`, `trainer`, `ci`, `training_utils`, `recipe`, `ray`, `worker`, `single_controller`, `misc`, `perf`, `model`, `algo`, `env`, `tool`, `ckpt`, `doc`, `data`, `cfg`, `reward`, `diffusion`, `omni`, `tests`, `docker`
  - If this PR involves multiple modules, separate them with `,` like `[diffusion, doc]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][diffusion, fsdp] feat: new rollout scheduler`

### Test

> For changes that can not be tested by CI (e.g., algorithm implementation, new model support), validate by experiment(s) and show results like training curve plots, evaluation results, etc.

### API and Usage Example

> Demonstrate how the API changes if any, and provide usage example(s) if possible.

```python
# Add code snippet or script demonstrating how to use this
```

### Design & Code Changes

> Demonstrate the high-level design if this PR is complex, and list the specific changes.

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [x] Read the [Contribute Guide](https://github.com/verl-project/verl-omni/blob/main/CONTRIBUTING.md).
- [x] Apply [pre-commit checks](https://github.com/verl-project/verl-omni/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [x] Add / Update the documentation.
- [x] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/verl-project/verl-omni/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...

### CODE REVIEW

File: docs/start/sd35_drm_flow_grpo.md [hunk 1]
```diff
@@ -0,0 +1,139 @@
+(sd35_drm_flow_grpo)=
```
Comment (wtomin) [inline docs/start/sd35_drm_flow_grpo.md]:
```suggestion
```

**Comment by zhtmike** [comment]
nice work!

File: docs/start/sd35_drm_flow_grpo.md [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+`$WORKSPACE/data/pickscore_sfw/sd3`. PickScore-SFW supplies prompts only;
+DiNa-LRM remains the reward model.
+
+## Step 2: Start the DRM server
+
+Start the scorer from its separate `diffusion-rm` environment. This example
+reserves physical GPU 7 for the server:
+
+```bash
+cd ../diffusion-rm
```
Comment (wtomin) [inline docs/start/sd35_drm_flow_grpo.md]: Do you need to clone the `diffusion-rm` first?


```suggestion
cd ../
git clone https://github.com/HKUST-C4G/diffusion-rm.git
cd diffusion-rm
```

File: examples/flowgrpo_trainer/sd35/run_sd35_medium_drm_lora.sh [hunk 3]
```diff
@@ -0,0 +1,119 @@
+#!/usr/bin/env bash
```
Comment (wtomin) [inline examples/flowgrpo_trainer/sd35/run_sd35_medium_drm_lora.sh:1]: Please follow the naming convention that renames this file to `run_sd35_medium_drm_pickscore_lora.sh`

**Review by SamitHuang — COMMENTED** [review]
Nice work. Please update main page readme as well for this algorithm support (News for team-original algo, and Supported models section)

File: verl_omni/workers/rollout/vllm_rollout/vllm_omni_async_server.py [hunk 4]
```diff
@@ -170,6 +182,28 @@ async def run_server(self, args: argparse.Namespace):
...
+                    raise RuntimeError(
+                        "The pinned vLLM-Omni SD3 pipeline no longer exposes "
+                        "get_sd3_image_post_process_func; latent output is incompatible with this version."
+                    )
+
+                def get_latent_post_process_func(od_config):
+                    return _build_sd3_latent_aware_postprocess(postprocess_factory, od_config)
+
+                pipeline_sd3.get_sd3_image_post_process_func = get_latent_post_process_func
+
```
Comment (zhtmike) [inline verl_omni/workers/rollout/vllm_rollout/vllm_omni_async_server.py]: model specific implementation should not put in general lib.

File: docs/start/sd35_drm_flow_grpo.md (hunk 2 above)
Comment (GongyeLiu) [reply docs/start/sd35_drm_flow_grpo.md]: Yes. The reward computation is implemented in `diffusion-rm`, while this pr maintains the HTTP client interface in `verl-omni`.

File: verl_omni/workers/rollout/vllm_rollout/vllm_omni_async_server.py (hunk 4 above)
Comment (wtomin) [reply verl_omni/workers/rollout/vllm_rollout/vllm_omni_async_server.py]: I agree. I think you can move this patch to `verl_omni/pipelines/sd3_flow_grpo/`, for example, `StableDiffsuion3PipelineWithLogProb.get_sd3_image_post_process_fun=get_latent_post_process_func`.

File: verl_omni/utils/reward_score/latent_http_scorer_client.py [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+def _serialize_request(
+    solution_image: torch.Tensor,
+    extra_info: dict,
+    noise_level: float,
+    noise_seed: int | None,
+) -> bytes:
+    latents = extra_info.get("latents_clean")
+    if latents is None and isinstance(solution_image, torch.Tensor) and solution_image.shape[-3] == 16:
+        latents = solution_image
+
```
Comment (wtomin) [inline verl_omni/utils/reward_score/latent_http_scorer_client.py:56]: Please add unit test for this file.

File: verl_omni/workers/rollout/vllm_rollout/vllm_omni_async_server.py (hunk 4 above)
Comment (GongyeLiu) [reply verl_omni/workers/rollout/vllm_rollout/vllm_omni_async_server.py]: Thanks for the suggestion! This has been addressed in 979435c. The SD3-specific postprocessing override has been moved to `verl_omni/pipelines/sd3_flow_grpo/`

File: verl_omni/utils/reward_score/latent_http_scorer_client.py (hunk 5 above)
Comment (GongyeLiu) [reply verl_omni/utils/reward_score/latent_http_scorer_client.py:56]: Thanks for the suggestion. This has been added in d0e0411.

File: verl_omni/utils/reward_score/latent_http_scorer_client.py [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+        tensors["seeds"] = torch.tensor([noise_seed], dtype=torch.int64)
+    return save_tensors(tensors)
+
+
+async def _session() -> aiohttp.ClientSession:
+    session = getattr(compute_score, "_session", None)
+    if session is None or session.closed:
+        session = aiohttp.ClientSession(timeout=aiohttp.ClientTimeout(total=None))
+        compute_score._session = session
+    return session
```
[the remark this answers is not in the document]
Comment (GongyeLiu) [reply verl_omni/utils/reward_score/latent_http_scorer_client.py:82]: Resolved in c0e1451

File: verl_omni/utils/reward_score/latent_http_scorer_client.py [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+    max_retries: int = 2,
+    retry_backoff: float = 0.5,
+    **kwargs,
+) -> dict:
+    """Score one SD3.5 clean latent with an external DRM server.
+
+    The rollout must use ``output_type=latent`` or ``output_type=both`` so
+    ``extra_info`` contains ``latents_clean``. Prompt embeddings are reused
+    from the rollout and therefore remain exactly aligned with generation.
+    """
```
[the remark this answers is not in the document]
Comment (zhtmike) [reply verl_omni/utils/reward_score/latent_http_scorer_client.py:139]: pls update doc

**Comment by zhtmike** [comment]
pls fix ci as well

**Comment by wtomin** [comment]
@GongyeLiu Could install pre-commit and run `pre-commit run --all-files`?

File: verl_omni/utils/reward_score/latent_http_scorer_client.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (GongyeLiu) [reply verl_omni/utils/reward_score/latent_http_scorer_client.py:139]: Thanks. Done in 9f63d62

**Comment by zhtmike** [comment]
ci still fail~

**Review by zhtmike — APPROVED** [review]
Thanks for the Work!
